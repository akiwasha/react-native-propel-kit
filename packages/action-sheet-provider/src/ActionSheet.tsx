import ModalDialog, {ModalDialogHandle, ModalDialogProps} from '@mgcrea/react-native-modal-dialog';
import React, {forwardRef, ForwardRefRenderFunction, useMemo} from 'react';
import {
  ActionSheetIOSOptions,
  Dimensions,
  Platform,
  ScrollView,
  ScrollViewProps,
  StyleProp,
  StyleSheet,
  TextStyle,
  ViewStyle
} from 'react-native';
import ActionSheetOption from './components/ActionSheetOption';
import isUndefined from './utils/isUndefined';

const {height: WINDOW_HEIGHT} = Dimensions.get('window');
const IOS_OPACITY = 0.85; // @NOTE from native ActionSheet (though it uses blur)

export type ActionSheetProps = Pick<ScrollViewProps, 'scrollEnabled'> &
  Omit<ModalDialogProps, 'onConfirm' | 'confirmTitle' | 'confirmStyle'> &
  ActionSheetIOSOptions & {
    defaultStyles?: typeof defaultStyles;
    optionStyle?: StyleProp<TextStyle>;
    lastOptionExtraStyle?: StyleProp<TextStyle>;
    destructiveButtonColor?: string;
    onButtonPress: (buttonIndex: number) => void;
    onCancel: () => void;
  };

export const defaultProps = {
  backgroundColor: 'white',
  cancelTitle: 'Cancel',
  destructiveButtonColor: Platform.select({
    ios: '#ff3b30', // iOS.systemRed (@see https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/color/)
    android: '#f44336' // android.red500 (@see https://material.io/design/color/the-color-system.html)
  })!
};

const ActionSheetRenderFunction: ForwardRefRenderFunction<ModalDialogHandle, ActionSheetProps> = (
  {
    options,
    backgroundColor = defaultProps.backgroundColor,
    optionStyle: propOptionStyle,
    lastOptionExtraStyle: propLastOptionExtraStyle,
    cancelButtonIndex,
    destructiveButtonIndex,
    destructiveButtonColor = defaultProps.destructiveButtonColor,
    onButtonPress,
    scrollEnabled: propScrollEnabled,
    defaultStyles: propDefaultStyles = defaultStyles,
    ...otherModalDialogProps
  },
  ref
) => {
  const cancelTitle = !isUndefined(cancelButtonIndex) && options[cancelButtonIndex] ? options[cancelButtonIndex] : '';
  const scrollEnabled = !isUndefined(cancelButtonIndex) ? options.length > 6 : propScrollEnabled;

  const optionStyle = useMemo<ViewStyle>(
    () => StyleSheet.flatten([propDefaultStyles.option, {backgroundColor}, propOptionStyle]),
    [propDefaultStyles, backgroundColor, propOptionStyle]
  );
  const lastOptionExtraStyle = useMemo<ViewStyle>(
    () => StyleSheet.flatten([propDefaultStyles.lastOptionExtra, propLastOptionExtraStyle]),
    [propDefaultStyles, propLastOptionExtraStyle]
  );

  return (
    <ModalDialog
      ref={ref}
      defaultStyles={propDefaultStyles}
      backgroundColor={backgroundColor}
      cancelTitle={cancelTitle}
      bodyStyle={{backgroundColor: 'transparent'}}
      {...otherModalDialogProps}>
      <ScrollView scrollEnabled={scrollEnabled}>
        {options.map((buttonTitle, buttonIndex) => {
          // Skip cancel button
          if (!isUndefined(cancelButtonIndex) && buttonIndex === cancelButtonIndex) {
            return null;
          }

          const color: string =
            destructiveButtonIndex && buttonIndex === destructiveButtonIndex ? destructiveButtonColor : '';
          return (
            <ActionSheetOption
              style={[optionStyle as TextStyle].concat(
                color ? [{color}] : [],
                buttonIndex === options.length - 1 ? [lastOptionExtraStyle] : []
              )}
              onPress={() => onButtonPress(buttonIndex)}
              title={buttonTitle}
              key={`${buttonIndex}`} // eslint-disable-line react/no-array-index-key
            />
          );
        })}
      </ScrollView>
    </ModalDialog>
  );
};

export const ActionSheet = forwardRef(ActionSheetRenderFunction);

export const defaultStyles = {
  modal: Platform.select<ViewStyle>({
    ios: {flex: 1, justifyContent: 'flex-end'},
    android: {flex: 1, justifyContent: 'center'}
  })!,
  container: Platform.select<ViewStyle>({
    ios: {
      flexDirection: 'column',
      alignItems: 'stretch',
      marginHorizontal: 8,
      marginBottom: 8,
      maxHeight: WINDOW_HEIGHT - 8
    },
    android: {
      flexDirection: 'column',
      alignItems: 'stretch',
      // marginHorizontal: 82, // timepicker/spinner
      margin: 24,
      maxHeight: WINDOW_HEIGHT - 24 * 2,
      elevation: 4
    }
  })!,
  header: Platform.select<ViewStyle>({
    ios: {
      alignItems: 'center',
      padding: 16,
      borderTopLeftRadius: 12,
      borderTopRightRadius: 12,
      opacity: IOS_OPACITY
    },
    android: {alignItems: 'flex-start', paddingVertical: 24, paddingHorizontal: 32}
  })!,
  body: Platform.select<ViewStyle>({
    ios: {
      flexShrink: 1,
      flexDirection: 'column',
      backgroundColor: 'transparent',
      opacity: IOS_OPACITY,
      borderBottomStartRadius: 12,
      borderBottomEndRadius: 12
    },
    android: {
      flexShrink: 1,
      flexDirection: 'column',
      backgroundColor: 'transparent'
    }
  })!,
  footer: Platform.select<TextStyle>({
    ios: {}, // @NOTE we don't have a footer on iOS
    android: {
      flexDirection: 'row',
      justifyContent: 'flex-end',
      borderBottomLeftRadius: 2,
      borderBottomRightRadius: 2,
      padding: 8
    }
  })!,
  title: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 14, fontWeight: '500', textAlign: 'center', color: '#888'},
    android: {
      paddingBottom: 12,
      fontSize: 20,
      fontWeight: '500',
      textAlign: 'left',
      color: '#333'
    }
  })!,
  message: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 13, fontWeight: '400', textAlign: 'center', color: '#888'},
    android: {paddingBottom: 12, fontSize: 18, fontWeight: '400', color: '#666'}
  })!,
  cancel: Platform.select<TextStyle>({
    ios: {marginTop: 8, borderRadius: 12, fontWeight: '600'},
    android: {}
  })!,
  confirm: Platform.select<TextStyle>({
    ios: {}, // @NOTE we don't have a confirm on iOS
    android: {}
  })!,
  option: Platform.select<TextStyle>({
    ios: {
      marginTop: StyleSheet.hairlineWidth
    },
    android: {
      marginTop: StyleSheet.hairlineWidth
    }
  })!,
  lastOptionExtra: Platform.select<TextStyle>({
    ios: {
      borderBottomLeftRadius: 12,
      borderBottomRightRadius: 12
    },
    android: {
      marginBottom: StyleSheet.hairlineWidth
    }
  })!
};
