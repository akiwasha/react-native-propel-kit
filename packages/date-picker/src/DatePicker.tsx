import {InputButton, InputButtonProps} from '@mgcrea/react-native-button';
import ModalDialog, {ModalDialogHandle, ModalDialogProps} from '@mgcrea/react-native-modal-dialog';
import DateTimePicker, {AndroidNativeProps, IOSNativeProps} from '@react-native-community/datetimepicker';
import React, {
  ElementType,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from 'react';
import {Platform, View} from 'react-native';
import {
  asLocaleDate,
  asStartOfDate,
  asUTCDate,
  DatePickerAndroid,
  defaultLabelExtractor,
  getNavigatorLocale,
  isUndefined,
  LabelExtractorOptions,
  openAndroidDatePicker,
  trimDate
} from './utils';

// const CURRENT_YEAR = new Date().getFullYear();
// const FIRST_DAY_OF_YEAR = new Date(Date.UTC(CURRENT_YEAR, 0, 1));
const START_OF_TODAY = asStartOfDate(new Date(), 'day');

export type DatePickerHandle = {
  focus: InputButtonProps['onFocus'];
};

export type DatePickerProps = Pick<ModalDialogProps, 'title' | 'confirmTitle' | 'cancelTitle'> &
  Pick<IOSNativeProps, 'mode' | 'display' | 'locale'> & {
    androidDisplay?: AndroidNativeProps['display'];
    initialValue?: Date;
    InputButtonComponent?: ElementType<InputButtonProps>;
    labelExtractor?: (value: Date, options: LabelExtractorOptions) => string;
    onChange?: (value: Date) => void;
    placeholder?: string;
    onSubmitEditing?: (value: Date) => void; // Callback that is called when the text input's submit button is pressed.
    value?: Date;
    utc?: boolean;
    trim?: boolean;
    [s: string]: unknown; // otherInputButtonProps
  };

export const defaultProps: Required<
  Pick<
    DatePickerProps,
    | 'androidDisplay'
    | 'initialValue'
    | 'InputButtonComponent'
    | 'labelExtractor'
    | 'locale'
    | 'mode'
    | 'display'
    | 'utc'
    | 'trim'
  >
> = {
  androidDisplay: 'default',
  initialValue: START_OF_TODAY,
  InputButtonComponent: InputButton,
  labelExtractor: defaultLabelExtractor,
  locale: getNavigatorLocale(),
  mode: 'date',
  display: 'inline',
  utc: false,
  trim: false
};

export const DatePicker = forwardRef<DatePickerHandle, DatePickerProps>(
  (
    {
      // ModalDialog props
      cancelTitle,
      confirmTitle,
      title,
      // InputButton props
      placeholder,
      // DatePicker props
      mode = defaultProps.mode,
      InputButtonComponent = defaultProps.InputButtonComponent,
      androidDisplay = defaultProps.androidDisplay,
      initialValue = defaultProps.initialValue,
      labelExtractor = defaultProps.labelExtractor,
      locale = defaultProps.locale,
      onChange: propOnChange,
      value: propValue,
      utc = defaultProps.utc,
      trim = defaultProps.trim,
      display = defaultProps.display,
      onSubmitEditing,
      ...otherInputButtonProps
    },
    ref
  ) => {
    const value = propValue || initialValue;
    const [modalValue, setModalValue] = useState<Date>(value);

    // Track parent propValue controlled updates
    useEffect(() => {
      const nextValue = utc ? asLocaleDate(value) : value;
      setModalValue(nextValue);
    }, [value, utc]);

    // Support both controlled/uncontrolled usage
    const [localValue, setLocalValue] = useState<Date>(initialValue);
    const inputValue = !isUndefined(propValue) ? propValue : localValue;

    // Lazily compute displayed label
    const labelValue = useMemo<string>(() => {
      if (isUndefined(inputValue)) {
        return '';
      }
      if (labelExtractor) {
        return labelExtractor(inputValue, {mode, locale, utc});
      }
      return `${inputValue}`;
    }, [labelExtractor, inputValue, mode, locale, utc]);

    const computeDate = (date: Date = initialValue) => {
      let nextValue = date;
      // console.warn({nextValue});
      if (trim) {
        nextValue = trimDate(nextValue, mode);
        // console.warn({trim: nextValue});
      }
      if (utc) {
        nextValue = asUTCDate(nextValue);
        // console.warn({utc: nextValue});
      }
      return nextValue;
    };

    // Propagate changed value
    const onConfirm = useCallback(
      (value?: Date) => {
        // @NOTE android uses direct calls while ios relies on a modalValue
        const nextValue = computeDate(Platform.select({ios: modalValue, android: value}));
        if (propOnChange) {
          propOnChange(nextValue);
        }
        if (onSubmitEditing) {
          // @NOTE prevent swallowing siblings focus events
          setTimeout(() => {
            onSubmitEditing(nextValue);
          });
        }
        // Support uncontrolled usage
        if (isUndefined(propValue)) {
          setLocalValue(nextValue);
        }
      },
      [mode, utc, propValue, modalValue, propOnChange, onSubmitEditing]
    );

    // Reset modalValue to the proper value
    const onCancel = useCallback(() => {
      setModalValue(inputValue || initialValue);
    }, [inputValue]);

    // Open the modal when user touches input
    const modalDialogRef = useRef<ModalDialogHandle>(null);
    const focus = Platform.select<InputButtonProps['onFocus']>({
      android: useCallback(async () => {
        const {action, value: nextValue} = await openAndroidDatePicker(mode, {
          value: inputValue,
          display: androidDisplay,
          utc
        });
        if (action !== DatePickerAndroid.dismissedAction) {
          onConfirm(computeDate(nextValue));
        }
      }, [inputValue, mode, onConfirm]),
      default: useCallback(() => {
        if (modalDialogRef.current) {
          modalDialogRef.current.show();
        }
      }, [])
    });
    // Expose API via an imperative handle
    useImperativeHandle<DatePickerHandle, DatePickerHandle>(ref, () => ({focus}), [focus]);

    return (
      <>
        <InputButtonComponent onFocus={focus} placeholder={placeholder} value={labelValue} {...otherInputButtonProps} />
        {Platform.OS === 'ios' ? (
          <ModalDialog
            ref={modalDialogRef}
            title={title || placeholder}
            onConfirm={onConfirm}
            onCancel={onCancel}
            confirmTitle={confirmTitle}
            cancelTitle={cancelTitle}
          >
            <View style={{height: mode === 'datetime' ? 417.5 : 368, width: '100%'}}>
              <DateTimePicker
                value={modalValue}
                mode={mode}
                style={{flex: 1}}
                display={display}
                onChange={(_event, date) => {
                  if (date) {
                    setModalValue(date);
                  }
                }}
              />
            </View>
          </ModalDialog>
        ) : null}
      </>
    );
  }
);

// 368
//417.5
