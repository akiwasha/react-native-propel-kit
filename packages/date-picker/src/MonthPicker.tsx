import {InputButton, InputButtonProps} from '@mgcrea/react-native-button';
import ModalDialog, {ModalDialogHandle, ModalDialogProps} from '@mgcrea/react-native-modal-dialog';
import {Picker} from '@react-native-picker/picker';
import React, {
  ElementType,
  forwardRef,
  ReactNode,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from 'react';
import {DatePickerIOSProps} from 'react-native';
import {
  asLocaleDate,
  asUTCDate,
  defaultLabelExtractor,
  getNavigatorLocale,
  isUndefined,
  LabelExtractorOptions,
  localizedMonths,
  trimDate
} from './utils';

const CURRENT_YEAR = new Date().getFullYear();
const CURRENT_MONTH = new Date().getMonth();
const START_OF_MONTH = new Date(Date.UTC(CURRENT_YEAR, CURRENT_MONTH, 1));

type MonthPickerValue = Date;

export type MonthPickerHandle = {
  focus: () => void;
};

export type MonthPickerProps = Pick<ModalDialogProps, 'title' | 'confirmTitle' | 'cancelTitle'> &
  Pick<DatePickerIOSProps, 'locale'> & {
    children?: ReactNode;
    initialValue?: MonthPickerValue;
    InputButtonComponent?: ElementType<InputButtonProps>;
    labelExtractor?: (value: Date, options: LabelExtractorOptions) => string;
    onChange?: (value: MonthPickerValue) => void;
    onSubmitEditing?: (value: MonthPickerValue) => void;
    placeholder?: string;
    minYear?: number;
    maxYear?: number;
    value?: MonthPickerValue;
    utc?: boolean;
    trim?: boolean;
    [s: string]: unknown; // otherInputButtonProps
  };

export const defaultProps: Required<
  Pick<MonthPickerProps, 'initialValue' | 'InputButtonComponent' | 'labelExtractor' | 'locale' | 'utc' | 'trim'>
> = {
  initialValue: START_OF_MONTH,
  InputButtonComponent: InputButton,
  labelExtractor: defaultLabelExtractor,
  locale: getNavigatorLocale(),
  utc: false,
  trim: false
};

export const MonthPicker = forwardRef<MonthPickerHandle, MonthPickerProps>(
  (
    {
      // ModalDialog props
      cancelTitle,
      confirmTitle,
      title,
      // InputButton props
      placeholder,
      // MonthPicker props
      initialValue = defaultProps.initialValue,
      InputButtonComponent = defaultProps.InputButtonComponent,
      labelExtractor = defaultProps.labelExtractor,
      locale = defaultProps.locale,
      utc = defaultProps.utc,
      onChange: propOnChange,
      onSubmitEditing,
      value: propValue,
      trim,
      minYear = CURRENT_YEAR - 100,
      maxYear = CURRENT_YEAR + 10,
      ...otherInputButtonProps
    },
    ref
  ) => {
    const value = propValue || initialValue;
    const [modalMonthValue, setModalMonthValue] = useState<number>(initialValue.getMonth());
    const [modalYearValue, setModalYearValue] = useState<number>(initialValue.getFullYear());

    // Track parent propValue controlled updates
    useEffect(() => {
      const nextValue = utc ? asLocaleDate(value) : value;
      setModalMonthValue(nextValue.getMonth());
      setModalYearValue(nextValue.getFullYear());
    }, [value]);

    // Support both controlled/uncontrolled usage
    const [localValue, setLocalValue] = useState<MonthPickerValue>(initialValue);
    const inputValue = useMemo(() => (!isUndefined(propValue) ? propValue : localValue), [propValue, localValue]);

    // Lazily compute displayed label
    const labelValue = useMemo<string>(() => {
      if (isUndefined(inputValue)) {
        return '';
      }
      if (labelExtractor) {
        return labelExtractor(inputValue, {mode: 'month', locale, utc});
      }
      return `${inputValue}`;
    }, [labelExtractor, inputValue, locale]);

    const computeDate = (modalYearValue: number, modalMonthValue: number) => {
      let nextValue = new Date(
        modalYearValue,
        modalMonthValue,
        value.getDate(),
        value.getHours(),
        value.getMinutes(),
        value.getSeconds(),
        value.getMilliseconds()
      );
      console.warn({nextValue});
      if (trim) {
        nextValue = trimDate(nextValue, 'month');
      }
      if (utc) {
        nextValue = asUTCDate(nextValue);
      }
      return nextValue;
    };

    // Propagate changed value
    const onConfirm = useCallback(() => {
      const nextValue = computeDate(modalYearValue, modalMonthValue);
      if (propOnChange) {
        propOnChange(nextValue);
      }
      if (onSubmitEditing) {
        // @NOTE Add a timeout to prevent swallowing siblings focus events
        setTimeout(() => {
          onSubmitEditing(nextValue);
        });
      }
      // Support uncontrolled usage
      setLocalValue(nextValue);
    }, [modalYearValue, modalMonthValue, onSubmitEditing, utc, propOnChange]);

    // Reset modalValue to the proper value
    const onCancel = useCallback(() => {
      setModalMonthValue((inputValue || initialValue).getMonth());
      setModalYearValue((inputValue || initialValue).getFullYear());
    }, [inputValue, initialValue]);

    // Lazily compute year options
    const yearOptions = useMemo(() => {
      const options = [];
      for (let i = minYear; i <= maxYear; i += 1) {
        options.push({label: `${i}`, value: i});
      }
      return options;
    }, [minYear, maxYear]);

    // Open the modal when user touches input
    const modalDialogRef = useRef<ModalDialogHandle>(null);
    const focus = useCallback(() => {
      if (modalDialogRef.current) {
        modalDialogRef.current.show();
      }
    }, []);
    // Expose API via an imperative handle
    useImperativeHandle(ref, () => ({focus}), [focus]);

    // Track modal value updates
    const onMonthValueChange = useCallback((value) => setModalMonthValue(value), []);
    const onYearValueChange = useCallback((value) => setModalYearValue(value), []);

    // Compute localized values
    const pickerValues = useMemo(() => {
      const lang = locale.split('-')[0];
      const months = localizedMonths[lang] || localizedMonths.en;
      return months.map((label, value) => ({label, value}));
    }, [locale]);

    return (
      <>
        <InputButtonComponent onFocus={focus} placeholder={placeholder} value={labelValue} {...otherInputButtonProps} />
        <ModalDialog
          ref={modalDialogRef}
          title={title || placeholder}
          onConfirm={onConfirm}
          onCancel={onCancel}
          confirmTitle={confirmTitle}
          cancelTitle={cancelTitle}
          bodyStyle={{flexDirection: 'row'}}
        >
          <Picker
            style={{flexBasis: '50%', flexGrow: 0, flexShrink: 0}}
            onValueChange={onMonthValueChange}
            selectedValue={modalMonthValue}
          >
            {pickerValues.map(({label, value}) => (
              <Picker.Item key={value} label={label} value={value} />
            ))}
          </Picker>
          <Picker
            style={{flexBasis: '50%', flexGrow: 0, flexShrink: 0}}
            onValueChange={onYearValueChange}
            selectedValue={modalYearValue}
          >
            {yearOptions.map(({label, value}) => (
              <Picker.Item key={value} label={label} value={value} />
            ))}
          </Picker>
        </ModalDialog>
      </>
    );
  }
);

export const MonthPickerItem = Picker.Item;
