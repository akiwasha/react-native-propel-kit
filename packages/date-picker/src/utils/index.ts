export * from './date';
export * from './defaultLabelExtractor';
export * from './getNavigatorLocale';
export * from './isUndefined';
export * from './localizedMonths';
export * from './openAndroidDatePicker';
