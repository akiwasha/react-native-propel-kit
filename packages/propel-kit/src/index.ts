export * from '@mgcrea/react-native-action-sheet-provider';
export * from '@mgcrea/react-native-backdrop-provider';
export * from '@mgcrea/react-native-button';
export * from '@mgcrea/react-native-date-picker';
export * from '@mgcrea/react-native-modal-dialog';
export * from '@mgcrea/react-native-select';
