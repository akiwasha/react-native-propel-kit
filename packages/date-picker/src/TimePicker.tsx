import React, {forwardRef} from 'react';
import {DatePicker, DatePickerHandle, DatePickerProps} from './DatePicker';

export type TimePickerProps = DatePickerProps;
export type TimePickerHandle = DatePickerHandle;

export const TimePicker = forwardRef<TimePickerHandle, TimePickerProps>(
  ({mode = 'time', display = 'spinner', ...otherProps}, ref) => (
    <DatePicker ref={ref} mode={mode} display={display} {...otherProps} />
  )
);
