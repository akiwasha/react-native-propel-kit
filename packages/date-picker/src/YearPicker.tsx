import {InputButton} from '@mgcrea/react-native-button';
import Select, {SelectProps} from '@mgcrea/react-native-select';
import React, {forwardRef, useMemo} from 'react';
import {getNavigatorLocale, LabelExtractorOptions} from './utils';

const CURRENT_YEAR = new Date().getFullYear();

export type YearPickerHandle = {
  focus: () => void;
};

export type YearPickerProps = SelectProps & {
  onChange?: (value: number) => void;
  labelExtractor?: (value: number, options: LabelExtractorOptions) => string;
  minValue?: number;
  maxValue?: number;
  locale?: string;
};

export const defaultProps = {
  initialValue: CURRENT_YEAR,
  InputButtonComponent: InputButton,
  labelExtractor: (value: number) => `${value}`,
  locale: getNavigatorLocale(),
  minValue: CURRENT_YEAR - 50,
  maxValue: CURRENT_YEAR + 20
};

export const YearPicker = forwardRef<YearPickerHandle, YearPickerProps>(
  (
    {
      initialValue: propInitialValue = defaultProps.initialValue,
      InputButtonComponent = defaultProps.InputButtonComponent,
      labelExtractor = defaultProps.labelExtractor,
      locale = defaultProps.locale,
      minValue = defaultProps.minValue,
      maxValue = defaultProps.maxValue,
      ...otherSelectProps
    },
    ref
  ) => {
    const yearOptions = useMemo(() => {
      const options = [];
      for (let value = minValue; value <= maxValue; value += 1) {
        options.push({label: labelExtractor ? labelExtractor(value, {mode: 'year', locale}) : `${value}`, value});
      }
      return options;
    }, [locale, labelExtractor, minValue, maxValue]);
    return (
      <Select
        InputButtonComponent={InputButtonComponent}
        initialValue={propInitialValue}
        ref={ref}
        {...otherSelectProps}
      >
        {yearOptions.map(({label, value}) => (
          <Select.Item key={value} label={label} value={value} />
        ))}
      </Select>
    );
  }
);
