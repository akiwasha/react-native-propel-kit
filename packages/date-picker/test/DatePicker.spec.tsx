import React from 'react';
import {fireEvent, render} from 'react-native-testing-library';
import {DatePicker} from '../src';

const DATE = new Date('2021-02-03T08:30:00.000Z');
const DATE_2 = new Date('2021-02-04T08:30:00.000Z');

describe('<DatePicker>', () => {
  it('renders correctly without props', () => {
    const component = render(<DatePicker />);
    expect(component.toJSON()).toMatchSnapshot();
  });
  describe('should support a `value` prop', () => {
    it('and render it', () => {
      const props = {value: DATE};
      const {getByText} = render(<DatePicker {...props} />);
      expect(getByText(props.value.toLocaleDateString())).toBeDefined();
    });
    it('and track parent updates', () => {
      const props = {value: DATE};
      const {getByText, update} = render(<DatePicker {...props} />);
      props.value = DATE_2;
      update(<DatePicker {...props} />);
      expect(getByText(props.value.toLocaleDateString())).toBeDefined();
    });
  });
  it('properly call onPress correctly', () => {
    const props = {value: DATE, onPress: jest.fn()};
    const {getByText} = render(<DatePicker {...props} />);
    fireEvent.press(getByText(props.value.toLocaleDateString()));
    expect(props.onPress).toBeCalledTimes(1);
    expect(props.onPress).toBeCalledWith(undefined); // ?
  });
});
