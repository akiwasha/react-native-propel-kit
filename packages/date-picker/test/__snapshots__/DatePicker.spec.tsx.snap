// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker> renders correctly without props 1`] = `
Array [
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 0,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "flex-start",
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "fontSize": 14,
          }
        }
      >
        11/9/2021
      </Text>
    </View>
  </View>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
          undefined,
        ]
      }
    >
      <RCTSafeAreaView
        accessible={true}
        emulateUnlessSupported={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "stretch",
              "flexDirection": "column",
              "marginBottom": 8,
              "marginHorizontal": 8,
              "maxHeight": 1326,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "flexShrink": 1,
              },
              Object {
                "borderTopLeftRadius": 12,
                "borderTopRightRadius": 12,
              },
              Object {
                "backgroundColor": "white",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "height": 368,
                "width": "100%",
              }
            }
          >
            <RNDateTimePicker
              date={1636495200000}
              disabled={false}
              displayIOS="inline"
              mode="date"
              onChange={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "height": 368,
                  },
                  Object {
                    "flex": 1,
                  },
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "borderBottomLeftRadius": 12,
                "borderBottomRightRadius": 12,
              },
              Object {
                "backgroundColor": "white",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 0,
                "flexDirection": "row",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderBottomLeftRadius": 12,
                  "borderBottomRightRadius": 12,
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "fontWeight": "400",
                  "height": 56,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#007aff",
                    "fontSize": 20,
                    "fontWeight": "400",
                  }
                }
              >
                Confirm
              </Text>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flex": 0,
              "flexDirection": "row",
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderRadius": 12,
                "flexDirection": "row",
                "flexGrow": 1,
                "fontWeight": "600",
                "height": 56,
                "justifyContent": "center",
                "marginTop": 24,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#007aff",
                  "fontSize": 20,
                  "fontWeight": "600",
                }
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </RCTSafeAreaView>
    </View>
  </Modal>,
]
`;
